### PR TITLE
Merge changes in the development branch

### PR DESCRIPTION
**Pull request by onurulgen**

### CODE REVIEW

**Comment by ericspod** [comment]
Do we know the issue with the tests? Also looking at the list of code analysis errors many are pure stylistic or about if-conditions related to the verbose code. Duplicate if-conditions can be condensed to one block but there's a number of things flagged that aren't duplicates so these checks maybe should be disabled.

**Comment by onurulgen** [comment]
> Do we know the issue with the tests? Also looking at the list of code analysis errors many are pure stylistic or about if-conditions related to the verbose code. Duplicate if-conditions can be condensed to one block but there's a number of things flagged that aren't duplicates so these checks maybe should be disabled.

No error with the tests, it just can't find the macro definition even though it can find the previous macro definition in the CemrgCommandLineTest.cpp. That's weird for sure. 

I can disable the style checking if it's requested.

**Comment by alonsoJASL** [comment]
I'm still sifting through the errors in https://github.com/CemrgDevelopers/CemrgApp/pull/53#issuecomment-893425407 . But I can say right now that anytime you find a 
```c++
if(verbose)
    MITK_INFO << "message";
``` 

You can change it for 
```c++
MITK_INFO(verbose) << "message"; 
```
that should get rid of a lot of those messages. I'll check the others and comment on them.

**Comment by ericspod** [comment]
The test job does indicate it failed though, is that not a concern?

**Comment by onurulgen** [comment]
Now, I've fixed most of the code analysis problems, only a few important ones are remaining. Please look into them.

**Comment by ericspod** [comment]
Many of the code issues above relate to some variable set a static value then used in a condition, I guess these are placeholder for setting the values from a GUI element? Others are about variable shadowing so the variable being flagged should be renamed. Those relating to unsafe types I'm not sure exactly what to do, those that don't have constructors but should probably do need constructors. Other things like using prefix ++/-- are clear about what to change. 

 CemrgApp/CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp Lines 444 the complaint is about 1.0 and 1.5 being cast to unsigned long which 1 in both cases, this should be changed to just be one if that was the intent of the ternary operator or changed to not cast to unsigned long.

**Review by alonsoJASL — DISMISSED** [review]
All looks good 👍 

Thanks @onurulgen for an amazing work.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    needs: [Variables]
+    steps:
+    - name: Uninstall NSIS
+      run: |
+        & 'C:/Program Files (x86)/NSIS/uninst-nsis.exe' /S
+      
+    - name: Download precompiled Build folder
+      working-directory: c:\
+      run: |
+        c:\msys64\usr\bin\wget.exe -q -O Build.zip https://emckclac-my.sharepoint.com/:u:/g/personal/k2143219_kcl_ac_uk/EUFa_AWPBI1OnnQwXA_ilu8Bq38U6OjiR4XKDcxDKo5wnw?download=1
```
Comment (OrodRazeghi) [inline .github/workflows/build.yml:147]: I think we should create an issue to remember to decouple these from personal accounts and find a more permanent place for them.

File: CemrgApp/Applications/MainApp/version.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+2.1.1
```
Comment (OrodRazeghi) [inline CemrgApp/Applications/MainApp/version.txt:1]: Does this get automatically generated?

File: CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp [hunk 3]
```diff
@@ -441,7 +439,7 @@ void CemrgAtriaClipper::ClipVeinsImage(std::vector<int> pickedSeedLabels, mitk::
 
         //Image Dilation
         BallType binaryBall;
-        binaryBall.SetRadius(manuals[i] == 1 ? static_cast<unsigned long>(1.0) : static_cast<unsigned long>(1.5));
+        binaryBall.SetRadius(manuals[i] == 1 ? 1 : 1.5); // manuals[i] == 1 ? 1 : 1.5
```
Comment (OrodRazeghi) [inline CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp]: Was this tested?

File: CemrgApp/Modules/CemrgAppModule/src/CemrgScar3D.cpp [hunk 4]
```diff
@@ -162,48 +158,46 @@ mitk::Surface::Pointer CemrgScar3D::Scar3D(std::string directory, mitk::Image::P
...
-        int _SCAR_MIP = 1;
+        MITK_INFO(debugging) << "int _ONLY_POSITIVE_STDEVS = 1";
+        MITK_INFO(debugging) << "int _SCAR_AS_STANDARD_DEVIATION = 1";
+        MITK_INFO(debugging) << "int _SCAR_MIP = 1";
         /**
          * @brief end
          */
 
-        if (_ONLY_POSITIVE_STDEVS == 1 && sdev < 0) sdev = 0;
+        // if (_ONLY_POSITIVE_STDEVS == 1 && sdev < 0) sdev = 0;
```
Comment (OrodRazeghi) [inline CemrgApp/Modules/CemrgAppModule/src/CemrgScar3D.cpp]: This needs a little bit of comments on why we are disabling them.

File: CemrgApp/Modules/CemrgAppModule/test/CemrgTestCommon.cpp [hunk 5]
```diff
@@ -16,35 +16,16 @@ PURPOSE.  See the above copyright notices for more information.
...
-
-  public:
-    void start(ctkPluginContext *context);
-    void stop(ctkPluginContext *context);
-
-  }; // kcl_cemrgapp_renderwindoweditor_Activator
-}
-
-#endif // kcl_cemrgapp_renderwindoweditor_Activator_h
+constexpr decltype(CemrgTestData::surfacePaths) CemrgTestData::surfacePaths;
```
Comment (OrodRazeghi) [inline CemrgApp/Modules/CemrgAppModule/test/CemrgTestCommon.cpp:31]: No sure what's happening here, is `renderwindoweditor` removed?

File: CemrgApp/Modules/CemrgAppModule/test/test.hpp [hunk 6]
```diff
@@ -16,26 +16,32 @@ PURPOSE.  See the above copyright notices for more information.
...
+// An example test class
+class TestSuite : public QObject {
 
-namespace mitk
-{
-  void kcl_cemrgapp_renderwindoweditor_Activator::start(ctkPluginContext *context)
-  {
-    BERRY_REGISTER_EXTENSION_CLASS(QmitkCemrgRenderWindowEditor, context)
-  }
+    Q_OBJECT
```
Comment (OrodRazeghi) [inline CemrgApp/Modules/CemrgAppModule/test/test.hpp:34]: same as the other comment about the renderwindoweditor

File: CemrgApp/Plugins/kcl.cemrgapp.wathca/src/internal/WallThicknessCalculationsView.cpp [hunk 7]
```diff
@@ -811,10 +815,10 @@ void WallThicknessCalculationsView::ThicknessCalculator() {
                 header[255] = '\n';
 
                 //Write to binary file
-                std::string path = (directory + mitk::IOUtil::GetDirectorySeparator() + "converted.inr").toStdString();
+                std::string path = (directory + "/converted.inr").toStdString();
                 ofstream myFile(path, ios::out | ios::binary);
                 myFile.write((char*)header, 256 * sizeof(char));
-                myFile.write((char*)&(*pv), dimensions * sizeof(uint8_t));
+                myFile.write((char*)pv, dimensions * sizeof(uint8_t));
```
Comment (OrodRazeghi) [inline CemrgApp/Plugins/kcl.cemrgapp.wathca/src/internal/WallThicknessCalculationsView.cpp:821]: Maybe worth double checking that there are the same after the change

**Review by OrodRazeghi — COMMENTED** [review]
@alonsoJASL almost everything looks good, only a few comments for you to have a look. Basically:

1. We need to create new issues for some of the topics mentioned in these comments, so we don't forget to come back to them in future.
2. Also, we have duplication of code in the `cmdapps`, which makes keeping track of changes in plugins and matching them with `cmdapps` difficult. I think we need a new issue for this too, to think about solutions in future.

File: CemrgApp/Applications/MainApp/version.txt (hunk 2 above)
Comment (alonsoJASL) [reply CemrgApp/Applications/MainApp/version.txt:1]: Yes, as I understand it, every time the github action gets launched, it calculates the current version and updates the file accordingly. Am I right, @onurulgen ? You can check it in line 107 of the [build.yml](https://github.com/CemrgDevelopers/CemrgApp/blob/development/.github/workflows/build.yml) file.

Comment (onurulgen) [reply CemrgApp/Applications/MainApp/version.txt:1]: Yes, that's correct. "version.txt" is automatically generated by the build system (github actions).

File: CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp (hunk 3 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp]: I tested it like that and the program worked for both manual and automatic clippers. We can play it safer and just do 
```c++
binaryBall.SetRadius(1); 
```
if you prefer

File: CemrgApp/Modules/CemrgAppModule/src/CemrgScar3D.cpp (hunk 4 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/src/CemrgScar3D.cpp]: I'll put something more useful. Do you want me to do it before the merge?

File: .github/workflows/build.yml (hunk 1 above)
Comment (alonsoJASL) [reply .github/workflows/build.yml:147]: Agreed, created #54

File: CemrgApp/Modules/CemrgAppModule/test/CemrgTestCommon.cpp (hunk 5 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/test/CemrgTestCommon.cpp:31]: Created #55

File: CemrgApp/Modules/CemrgAppModule/test/test.hpp (hunk 6 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/test/test.hpp:34]: #55

File: CemrgApp/Plugins/kcl.cemrgapp.wathca/src/internal/WallThicknessCalculationsView.cpp (hunk 7 above)
Comment (alonsoJASL) [reply CemrgApp/Plugins/kcl.cemrgapp.wathca/src/internal/WallThicknessCalculationsView.cpp:821]: I can't remember if I did this change. @onurulgen , do you remember this change? Does anyone know how can we check the functionality is the same?

Comment (alonsoJASL) [reply CemrgApp/Plugins/kcl.cemrgapp.wathca/src/internal/WallThicknessCalculationsView.cpp:821]: I think I changed this as part of the automated code-analysis checks from codecov.

Comment (onurulgen) [reply CemrgApp/Plugins/kcl.cemrgapp.wathca/src/internal/WallThicknessCalculationsView.cpp:821]: I changed it and it does the same thing, I tested it.

File: CemrgApp/Modules/CemrgAppModule/src/CemrgScar3D.cpp (hunk 4 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/src/CemrgScar3D.cpp]: Created #56

File: CemrgApp/Modules/CemrgAppModule/test/CemrgTestCommon.cpp (hunk 5 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/test/CemrgTestCommon.cpp:31]: For some reason, github is taking this `CemrgTestCommon.cpp` and comparing it to the recently deleted `kcl_cemrgapp_renderwindoweditor_Activator.h` . I don't know why github decided to do that.

File: CemrgApp/Modules/CemrgAppModule/test/test.hpp (hunk 6 above)
Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/test/test.hpp:34]: Same as the other comment. For some reason, github caught me deleting the `kcl_cemrgapp_renderwindoweditor_Activator.h` and put that change on this `test.hpp`. 

Not sure why, as they're in very different paths! Do you have any idea about this, @onurulgen ? 

For context, I deleted the renderwindow plugin we had in CemrgApp (commit 100fd8), but for some reason, github caught that deletion as a change in this test.hpp file

File: CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp (hunk 3 above)
Comment (OrodRazeghi) [reply CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp]: So, it was previously always set to 1; let's do that but if we keep seeing uncut veins then we should think about setting it to a larger value than 1.

Comment (alonsoJASL) [reply CemrgApp/Modules/CemrgAppModule/src/CemrgAtriaClipper.cpp]: Agreed. Commit 2b050da

**Review by OrodRazeghi — APPROVED** [review]

**Review by alonsoJASL — APPROVED** [review]
Thanks @OrodRazeghi for your feedback. I think we can merge this one for now. I created #54, #55, and #57 for the things that need to be addressed but not critically.

**Pull request merged**